src/System.Private.CoreLib/src/System/AppContext/AppContext.cs

```diff
@@ -37,7 +40,15 @@ public static string BaseDirectory
             {
                 // The value of APP_CONTEXT_BASE_DIRECTORY key has to be a string and it is not allowed to be any other type. 
                 // Otherwise the caller will get invalid cast exception
-                return (string)AppDomain.CurrentDomain.GetData("APP_CONTEXT_BASE_DIRECTORY") ?? AppDomain.CurrentDomain.BaseDirectory;
+                string baseDirectory = (string)AppDomain.CurrentDomain.GetData("APP_CONTEXT_BASE_DIRECTORY");
+                if (baseDirectory != null)
+                    return baseDirectory;
+
+                // Fallback path for hosts that do not set APP_CONTEXT_BASE_DIRECTORY explicitly
+                string directory = Path.GetDirectoryName(GetEntryAssembly()?.Location);
+                if (directory != null && !PathInternal.EndsInDirectorySeparator(directory))
+                    directory += Path.DirectorySeparatorChar;
+                return directory;
```

Comment:
Isn't this changing the return value?
If I'm reading the code correctly, before the change this would return the path to the current module (basically something like `GetDirectoryName(GetModuleFileName(null, ..)))`). That would mean it would return the directory where `coreclr.dll` lives.
After this change we're now returning the location of the application .exe (or its equivalent).

Reply:
Good catch. `GetDirectoryName(GetModuleFileName(null, ..))` would return path to the process .exe before, not path to coreclr.dll.

I agree that this is changing the behavior. I think we should be fine: It does not affect the official .NET Core host that always sets `APP_CONTEXT_BASE_DIRECTORY`, and presumably the new behavior is more sensible as fallback than what used to be there before. What do you think?

Reply:
I agree that for normal apps it doesn't matter since the host sets this.
Might be worth looking at #2128 though. Either we're basically fully fixing that bug now or we are reintroducing it a little bit:
* That bug has basically been fixed with the host change for all apps which use our host. So maybe we should close it
* The code you added which uses entry point assembly location - will it work with the symbolic links - basically will it work in the case the bug describes?

Reply:
Yes, this change will fix the bug for corerun.